sklearn/linear_model/_base.py

```diff
@@ -229,12 +229,12 @@ def _preprocess_data(X, y, fit_intercept, normalize=False, copy=True,
 
     if fit_intercept:
         if sp.issparse(X):
-            X_offset, X_var = mean_variance_axis(X, axis=0)
+            X_offset, X_var = mean_variance_axis(X, axis=0,
```

Comment:
I would find the following code a bit more readable since the scaling is the same for both sparse and dense (and we should remove the obsolete comment:

```python
        if sp.issparse(X):
            X_offset, X_var = mean_variance_axis(
                X, axis=0, weights=sample_weight
            )

            if not return_mean:
                X_offset[:] = X.dtype.type(0)
        else:
            X_offset, X_var, _ = _incremental_mean_and_var(
                X, last_mean=0., last_variance=0., last_sample_count=0.,
                sample_weight=sample_weight
            )

            X_offset = X_offset.astype(X.dtype)
            X -= X_offset

        X_var = X_var.astype(X.dtype)

        if normalize:
            X_var *= X.shape[0]
            X_scale = np.sqrt(X_var, X_var)
            del X_var
            X_scale[X_scale == 0] = 1
            if sp.issparse(X):
                inplace_column_scale(X, 1. / X_scale)
            else:
                X /= X_scale
        else:
            X_scale = np.ones(X.shape[1], dtype=X.dtype)
```

2 things might still be weird:

* `del X_var` should be useless since `X_scale` is still a reference on the same array (because of the inplace operation).
* if `return_mean` is linked to also not centering the data we could simplify a bit the code:

```python
        if sp.issparse(X):
            X_offset, X_var = mean_variance_axis(
                X, axis=0, weights=sample_weight
            )
        else:
            X_offset, X_var, _ = _incremental_mean_and_var(
                X, last_mean=0., last_variance=0., last_sample_count=0.,
                sample_weight=sample_weight
            )

        if not return_mean:
            X_offset[:] = X.dtype.type(0)
        else:
            X_offset = X_offset.astype(X.dtype)
            X -= X_offset
```